### DISCUSSION TITLE
Name conflict and use of `exclude_properties` in `Composite Columns`

### DISCUSSION THREAD

**Discussion by am1ter** [sqlalchemy/sqlalchemy#9869] [Usage Questions] [answered]
Hello everybody!

I have an issue with [Composite Column](https://docs.sqlalchemy.org/en/20/orm/composites.html) when I want to associate a single database column with my own single dataclass. This issue arises only when there is a name conflict between my Domain class and ORM class and I try to resolve this conflict using `exclude_properties` argument for `map_imperatively()` function. Please let me provide you some context:

I am developing my application using DDD approach. So, I have Domain model and I map it with the ORM model using the `map_imperatively()` method. In my domain model, I have Value Object attribute. I want to query the database and receive Domain-class instances (DomainUser) with my Value Object (Email) as one of its attributes.

```python
@dataclass
class DomainUser:
    email: Email
    name: str


@dataclass
class Email:
    email: str


class OrmUser(Base):
    name: Mapped[str]
    email: Mapped[str]


def init_orm_mappers() -> None:
    mapper_registry.map_imperatively(
        DomainUser,
        OrmUser, 
        properties={
            "email": composite(Email, OrmUser.email),
            "exclude_properties": [OrmUser.email],
        },
    )
```
As you can see, inside the `properties` argument of `map_imperatively()` function, I create an `email` attribute using the `composite()` function. If I rename it to `_email` and remove `exclude_properties` argument, everything will work well:
```python
>>> db.scalar(select(DomainUser).where(DomainUser.email = "test@test.com"))
DomainUser(email="test@test.com", name="Test", _email=Email(email="test@test.com"))
```
However, if I don't rename it, I will get an error:
```python
>>> db.scalar(select(DomainUser).where(DomainUser.email = Email("test@test.com")))
InvalidRequestError: Composite class Email is not a dataclass and does not define a __composite_values__() method; can't get state
```
Even if I try to use [Legacy Non-dataclasses](https://docs.sqlalchemy.org/en/20/orm/composites.html#using-legacy-non-dataclasses) and add the `__composite_values__()` double dash method to `Email` class I will get the same error.

I have analyzed the traceback, but it did not provide any useful information. I also couldn't find anything related in the documentation about Composite columns.

My conclusion is that it is impossible to use `exclude_properties` in a such case. However, I tried to do it because SQLAlchemy showed me an error that advised using it:
```python
WARNING: when configuring property 'email' on Mapper[DomainUser(user)], column 'email' conflicts with property '<Composite at 0x7ff873d71ed0; email>'.  To resolve this, map the column to the class under a different name in the 'properties' dictionary.  Or, to remove all awareness of the column entirely (including its availability as a foreign key), use the 'include_properties' or 'exclude_properties' mapper arguments to control specifically which table columns get mapped.
```

In the end, the only option I see right now is to rename the `email` attribute to something else to avoid the name conflict. Do you happen to know of any other approach to obtain an instance of `Email` inside my `DomainUser` after querying it from the database?

**Comment by zzzeek** [comment, marked as answer]
hi -

it looks like you are taking an existing Declarative class and then sending it into "map_imperatively".   This is not correct use of this function, so before we look into things with composites we have to figure out what this mapping is trying to do.

if you want to use imperative mappings, you would not use declarative base at all, you would set up `Table` objects.   

In this case, you could probably keep your OrmUser class, but anything passed into `map_imperatively()` would have to be in terms of `OrmUser.__table__`.

If I make just that change alone, everything works as in the runnable case below.    however I think the separation of dataclass from mapping leads to a poor result here, as the program cannot pass pep-484 typing.   The fact is that when you say something like ``select(DomainUser).where(DomainUser.email == Email("test@test.com"))``, that implies specific behaviors for `DomainUser` which are not implied by `@dataclass` alone, so I would argue the program is incorrect from a typing perspective (indeed, IMO all of imperative mapping is "wrong" as it can't be correctly typed, but we keep it because lots of users really like it).

work with the example below and you can modify it to show me further things that aren't working

```py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from typing import cast

from sqlalchemy import Column
from sqlalchemy import select
from sqlalchemy.orm import composite
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import registry


class Base(DeclarativeBase):
    pass


mapper_registry = registry()


@dataclass
class DomainUser:
    email: Email
    name: str


@dataclass
class Email:
    email: str


class OrmUser(Base):
    __tablename__ = "user"

    name: Mapped[str] = mapped_column(primary_key=True)
    email: Mapped[str]


def init_orm_mappers() -> None:
    mapper_registry.map_imperatively(
        DomainUser,
        OrmUser.__table__,
        properties={
            "email": composite(
                Email, cast(Column[Any], OrmUser.__table__.c.email)
            ),
            "exclude_properties": [OrmUser.__table__.c.email],
        },
    )
init_orm_mappers()

stmt = select(DomainUser).where(DomainUser.email == Email("test@test.com"))
print(stmt)



```

**Reply by am1ter** [reply]
Hello, @zzzeek! Thank you very much for your reply!
1. Yes, using ` OrmUser.__table__.c.email` instead of ` OrmUser.email` completely resolved my problem. 
2. You are correct that I used a declarative class for an imperative mapping function. I had assumed that it was an inappropriate way to use the mapping since the documentation did not provide information about it. However, in my opinion, the Declarative definition approach is clearer than using Declarative with imperative table or Imperative mapping. It has worked well so far, and this is the first issue I encountered with it.
3. In my real-world application, I want to use the `attrs` library instead of dataclasses for Value Objects. However, I currently need to add `__composite_values__` to it to force `map_imperatively()` works correctly. Could you please advice me how can use the `attrs` with `Composite Columns`?

**Reply by CaselIT** [reply]
How strange that `map_imperatively` did not raise when getting an model class. Maybe it should?

**Reply by zzzeek** [reply]
composite will work with any class that has a ``__composite_values__`` method, that's how it always worked.  the dataclass support, where the ``__composite_values__`` method is not required, is a new feature.  we don't offer this feature for an attrs class.

**Reply by am1ter** [reply]
> How strange that `map_imperatively` did not raise when getting an model class. Maybe it should?

I hope you won't :). Could you please enlighten me why you think it is bad idea to use the model class instead of the table in the case of `map_imperatively()`? Look at the sample above, it is so short and expressive. :)

**Reply by CaselIT** [reply]
Conceptually it makes no sense:
- you are creating a DomainUser class
- you are creating a mapped class OrmUser
- you are then taking domain user and transforming it into a mapped class that's equivalent to OrmUser

Can't you just use OrmUser? Why do you need to create two mapped classes that are the same?

**Reply by am1ter** [reply]
@CaselIT, @zzzeek I apologize for reaching out again, but I would greatly appreciate your advice on how to resolve another issue regarding `map_imperatively()`. After adding `__table__` to the properties I was able to successfully query the database and retrieve the correct objects. However, I encountered a problem when attempting to commit a new object to the database I get a blank values instead my Value Objects. Specifically, in the case mentioned earlier, the `db_session.new` identity map contains the correct object:
```python
DomainUser(name="test", email=Email(value="test@test.com"))
```
But when I call `db_session.commit()` an error is raised:
```python
NotNullViolationError: null value in column "email" of relation "user" violates not-null constraint
```
Yes, I applied non-blank constraint to `email` column of the database. Something works not as expected, but I do not understand what.

@CaselIT, Regarding my approach: I have two quite similar but separate models: the Domain model and the ORM model. I follow the principles of Domain-Driven Design, which emphasize the decoupling of domain logic from infrastructure concerns, such as database persistence. In my implementation, I encapsulate all of my business logic within rich domain models, while the ORM serves primarily as a means of storage. Within my project, I organize different components into specific layers: the database is described in the database layer, and the business logic resides within the domain layer, and so on.
Furthermore, I particularly appreciate SQLAlchemy's Declarative definition approach because it allows me to describe the database schema in a style that aligns closely with my domain models.

**Reply by CaselIT** [reply]
> @CaselIT, Regarding my approach: I have two quite similar but separate models: the Domain model and the ORM model. I follow the principles of Domain-Driven Design, which emphasize the decoupling of domain logic from infrastructure concerns, such as database persistence. In my implementation, I encapsulate all of my business logic within rich domain models, while the ORM serves primarily as a means of storage. Within my project, I organize different components into specific layers: the database is described in the database layer, and the business logic resides within the domain layer, and so on.
Furthermore, I particularly appreciate SQLAlchemy's Declarative definition approach because it allows me to describe the database schema in a style that aligns closely with my domain models

I'm not sure you are following that principle since you are transforming your domain model in the orm model. This is evidenced by the fact that you query the db using the domain model.

In your approach the orm model is useless, it's just a different way of defining a table.

I think what you want to so is something like what's described here https://docs.pydantic.dev/latest/usage/models/#orm-mode-aka-arbitrary-class-instances :
- you query the db using the orm model
- from the orm model you instantiate the domani model
- you make the changes you require
- from the domain model you create an updated instance of the orm model
- you save the changes to the db using the orm model

(Note that I'm not suggesting you do the above process, just that to keep the orm and the domain models separated that are roughly the steps required)
In your example the domain model _is_ the orm model, and you never instance the original OrmModel

**Comment by am1ter** [comment]
@zzzeek, @CaselIT hello once again!

In my previous discussion, I mentioned that I'm facing a problem when committing a new object to the database while using map_imperatively() with the exclude_properties attribute. I believe I understand where the problem lies, but I'm uncertain about how to fix it. Moreover, I'm not sure if I'm encountering the expected behavior. Please allow me to provide you some additional context:
```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from typing import cast

from sqlalchemy import Column
from sqlalchemy import select
from sqlalchemy.orm import composite
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import registry


class Base(DeclarativeBase):
    pass


mapper_registry = registry()


@dataclass
class DomainUser:
    name: str
    email: Email
    email: Password


@dataclass
class Email:
    value: str


@dataclass
class Password:
    value: str


class OrmUser(Base):
    __tablename__ = "user"

    name: Mapped[str] = mapped_column(primary_key=True)
    email: Mapped[str]
    password: Mapped[str]


def init_orm_mappers() -> None:
    mapper_registry.map_imperatively(
        DomainUser,
        OrmUser.__table__,
        properties={
            "email": composite(
                Email, cast(Column[Any], OrmUser.__table__.c.email)
            ),
            "password": composite(
                Password, cast(Column[Any], OrmUser.__table__.c.password)
            ),
            "exclude_properties": [OrmUser.__table__.c.email, OrmUser.__table__.c.password],
        },
    )
init_orm_mappers()

stmt = select(DomainUser).where(DomainUser.email == Email("test@test.com"))
new_user = DomainUser(name="test2", email=Email(value="test2@test2.com"), password=Password(value="test"))
db_session.add(new_user)
db_session.commit()
```
In the previous scenario, where I only had one Value Object (Email), the commit operation worked correctly. However, in the current case where I have two value objects (Email and Password), the mapping doesn't seem to work as expected. Only one Value Object (the last one defined in the mapping) is correctly mapped, while the other object appears as `null` during the `commit()` operation. Please take a look at the properties of the `new_user` instance:
```python
>>> new_user
>>> DomainUser(name="test", email=Email(value="test@test.com"), password=Password(value="test"))
>>> new_user.exclude_properties
>>> 'test'
```
During the `commit()` call, the following error is raised:
```python
NotNullViolationError: null value in column "email" of relation "user" violates not-null constraint
DETAIL:  Failing row contains (admin, null, test).
```

Moreover, when I inspect `Mapper` instance I see, that columns mapped incorrectly (email is mapped twice) despite `local_table` is configured correctly:
```python
>>> # It is incorrect
>>> list(mapper_registry.mappers)[0].c._all_columns
>>> [Column('name', String(), table=<user>, nullable=False), Column('email', String(), table=<user>, nullable=False), Column('email', String(), table=<user>, nullable=False)]
>>> # It is correct
>>> list(mapper_registry.mappers)[0].local_table.c._all_columns
>>> [Column('name', String(), table=<user>, nullable=False), Column('email', String(), table=<user>, nullable=False), Column('password', String(), table=<user>, nullable=False)]
```

If I exclude `exclude_properties` argument I get a correct mapping:
```python
>>> # It is correct
>>> list(mapper_registry.mappers)[0].c._all_columns
>>> [Column('name', String(), table=<user>, nullable=False), Column('email', String(), table=<user>, nullable=False), Column('password', String(), table=<user>, nullable=False)]
```

I do not understand why one attribute (password) mapped correctly and another one does not (email). I would really appreciate your advice on how to resolve this issue.

**Reply by CaselIT** [reply]
Hi

`exclude_properties` does not go inside `properties`. The issue you are having comes from it.
If you place it outside, things are still not working, since it does not find the column.
My suggestion is to use the `key` to indicate a different name for the email and password column, so that you don't have to ignore them.
```py
user = Table(
    "user",
    Base.metadata,
    Column("name", String, primary_key=True),
    Column("email", String, key='_email'),
    Column("password", String, key='_password'),
)

mapper_registry.map_imperatively(
    DomainUser,
    user,
    properties={
        "email": composite(Email, user.c._email),
        "password": composite(Password, user.c._password),
        
    },
)
```

I you want to continue using the declarative to define the table you can probably do the same doing
```py

class OrmUser(Base):
    __tablename__ = "user"

    name: Mapped[str] = mapped_column(primary_key=True)
    _email: Mapped[str] = mapped_column('email')
    _password: Mapped[str] = mapped_column('password')
```

**Comment by am1ter** [comment]
I have examined the source code of `Mapper.py` file using debugger and have discovered some rather peculiar behavior in the `Mapper._configure_property()` method. To me, it appears to be a bug, but I must admit that my proficiency in SQLAlchemy is not sufficient to definitively state that it is true. I will attempt to provide more details regarding my findings below.
```python
>>> # Line #2092 of sqlalchemy.orm.mapper (method _configure_property)
>>> # The execution stopped after the moment, when 2 columns (name and email) were already configurated
>>> self.columns._all_columns
>>> [Column('name', Strin...ble=False), Column('email', Stri...ble=False)]
>>> # New column creating during the run the following code line:
>>> col = self.persist_selectable.corresponding_column(prop.columns[0])
>>> # In the same time current prop.columns:
>>> prop.columns
>>> [Column('email', Stri...ble=False), Column('password', S...ble=False)]
>>> # So, the result of the defention of the line #2092 in that case is:
>>> col
>>> Column('email', String(), table=<user>, nullable=False)
```
And the code below does not change the situation. As the result we get 2 identical columns with `email`.

@zzzeek @CaselIT What do you think about it? If it is really seems like some kind of bug please tell me and I will report it as new Issue.

P.S. SQLAlchemy version is 2.0.15
